src/Microsoft.Windows.ComputeVirtualization/Container.cs

```diff
@@ -25,22 +24,49 @@ public class ProcessStartInfo
     /// </summary>
     public class Container : IDisposable
     {
-        private string _id;
+        private IntPtr _cs;
         private bool _killOnClose;
         private bool _killed;
 
-        internal Container(string id, bool terminateOnClose)
+        private HcsNotificationWatcher _watcher;
+
+        internal Container(string id, IntPtr computeSystem, bool terminateOnClose)
         {
-            _id = id;
             _killOnClose = terminateOnClose;
+            _cs = computeSystem;
+            _watcher = new HcsNotificationWatcher(
+                _cs,
+                HcsFunctions.HcsRegisterComputeSystemCallback,
+                HcsFunctions.HcsUnregisterComputeSystemCallback,
+                new HCS_NOTIFICATIONS[]{
+                    HCS_NOTIFICATIONS.HcsNotificationSystemExited,
+                    HCS_NOTIFICATIONS.HcsNotificationSystemCreateCompleted,
+                    HCS_NOTIFICATIONS.HcsNotificationSystemStartCompleted,
+                    HCS_NOTIFICATIONS.HcsNotificationServiceDisconnect
+                    }
+                );
+
+            _watcher.Wait(HCS_NOTIFICATIONS.HcsNotificationSystemCreateCompleted);
         }
 
         /// <summary>
         /// Starts the container.
         /// </summary>
         public void Start()
         {
-            HcsFunctions.StartComputeSystem(_id);
+            StartAsync();
+            WaitStart();
+        }
+
+        public void StartAsync()
```

Comment:
You should make these return `Tasks` from the beginning. Even though the low level stuff isn't `Task` based. IE:

```C#
public  Task StartAsync()
// ...
return Task.Factor.StartNew(() => { ... your code here ... })
// ...
```